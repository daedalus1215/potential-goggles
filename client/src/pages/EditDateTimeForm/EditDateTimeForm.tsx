import React, { useContext, useEffect } from 'react';
import cn from 'classnames';
import { Button, SaveButton } from '../../components';
import { DateTime, Task } from '../../interfaces';
import { formatMinsAndSecsForDisplay } from '../../utils';
import { minSecValidator, utcFormatValidator } from '../../utils/forms/validators';
import { Form, redirect, useLoaderData } from 'react-router-dom';

import styles from './EditDateTimeForm.module.css';
import { ExpandedContext } from '@/contexts/ExpandedContext';

interface EditDateTimeFormProp {
  taskId: string,
  dateTime: DateTime
}

const EditDateTimeForm: React.FC = () => {
  const { dateTime, taskId } = useLoaderData() as EditDateTimeFormProp;
  const {isExpanded} = useContext(ExpandedContext);
  const minsAndSecs = formatMinsAndSecsForDisplay(dateTime.time);

  if (!dateTime) {
    throw new Response("", {
      status: 404,
      statusText: "DateTime not found!",
    });
  }

  return (
    <div className={styles.editDateTimeForm}>
      <Form
        method='post'
        className={cn(styles.form, {[styles.notExpanded]: !isExpanded})}>
        <h2 className={styles.h2}>Edit Date Time</h2>
        <input hidden={true} name="taskId" value={taskId} readOnly />
        <input hidden={true} name="id" value={dateTime.id} readOnly />
        <input className={styles.input} type="text" name="date" defaultValue={dateTime.date} />
        <input className={styles.input} type="text" name="minutes" defaultValue={minsAndSecs} />
        <SaveButton />
      </Form>
    </div>
  );
};

export default EditDateTimeForm;
