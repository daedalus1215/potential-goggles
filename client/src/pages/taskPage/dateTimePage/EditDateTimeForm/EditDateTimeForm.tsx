import React, { useEffect } from 'react';
import cn from 'classnames';
import { Button } from '../../../../components';
import { DateTime, Task } from '../../../interfaces';
import { formatMinsAndSecsForDisplay } from '../../../../utils';

import styles from './EditDateTimeForm.module.css';
import { minSecValidator, utcFormatValidator } from '../../../../utils/forms/validators';
import { Form } from 'react-router-dom';

interface EditDateTimeFormProp {
  editDateTime: DateTime;
  task: Task;
  setIsShowingEditDateTimeForm: (setIsShowing: boolean) => void;
}

const EditDateTimeForm: React.FC<EditDateTimeFormProp> = ({ editDateTime, task, setIsShowingEditDateTimeForm }) => {
  const minsAndSecs = formatMinsAndSecsForDisplay(editDateTime.time);

  return (
    <Form
      method='post'
      //@TODO: tie in the action here!
      // action={}
      className="formEditDateTime"
    >
      <h2>Edit Date Time</h2>
      <input hidden={true} name="id" value={editDateTime.id} />
      <input hidden={true} name="taskId" value={task._id} />
      <div className={styles.field}>
        {/* <Field name="date" validate={utcFormatValidator}> */}
        <div className={styles.field}>
          <label htmlFor="date">Date</label>
          <input type="text" id="date" className={styles.input} />
        </div>
      </div>
      <div className={styles.field}>
        <label htmlFor="minutes">Minutes</label>
        <input type="text" id="minutes" className={cn(styles.input)} />
      </div>
      <Button type="submit" className={cn({ [styles.submit]: true })} value="Submit Form" />
    </Form>
  );
};

export default EditDateTimeForm;
