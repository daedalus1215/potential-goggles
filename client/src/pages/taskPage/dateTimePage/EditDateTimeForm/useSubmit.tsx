
interface SubmitProp {
  id: string;
  date: string;
  minutes: string;
}

const useSubmit = (taskId: String, setIsShowingEditDateTimeForm: (isShowing: boolean) => void): ((editDateTime: EditDateTimeInterface) => void) => {
  // const dispatch = useDispatch();
  return ({ id, date, minutes }: SubmitProp) => {
    const config = {
      body: {
        id,
        date,
        minutes,
      },
      taskId,
      dateTimeId: id,
    };

    //@TODO: Invoke an action and persist the Date Time
    // dispatch(putDateTime(config));
    setIsShowingEditDateTimeForm(false);
    // window.location.reload();
  };
};

export default useSubmit;
