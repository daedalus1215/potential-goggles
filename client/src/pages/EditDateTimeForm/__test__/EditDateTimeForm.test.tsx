import { render } from '@testing-library/react';
import '@testing-library/jest-dom'

// target
import EditDateTimeForm from '../EditDateTimeForm';

// mocks
jest.mock('@/hooks/useExpandedContext', () => ({
    default: () => ({ isExpanded: false })
}));
jest.mock("react-router-dom", () => ({
    ...(jest.requireActual("react-router-dom") as any),
    useLoaderData: () => ({ dateTime: { time: '123:23' }, taskId: 'randoTaskId' }),
    useSubmit: () => jest.fn(),
    Form: ({ children }: any) => <div data-testid="Form">({children})</div>,
}));
jest.mock('@/components', () => ({
    TopBar: ({ children }: any) => <div data-testid="TopBar">({children})</div>,
    SaveButton: () => <div data-testid="SaveButton">SaveButton</div>,
    BackButton: ({ path }: any) => <div data-testid="BackButton">{path}</div>
}));

describe('src/pages/tasks/TaskForm/dateTimeDetail/EditDateTimeForm/__test__/EditDateTimeForm.test.tsx', () => {
    describe('EditDateTimeForm', () => {
        it('should render', () => {
            // Arrange & Act
            const target = render(<EditDateTimeForm />);

            // Assert
            expect(target.queryByTestId("BackButton")).toBeInTheDocument();
            expect(target.queryByTestId("BackButton")).toHaveTextContent(`/task/randoTaskId/date-time`);
            expect(target).toMatchSnapshot();
        });
    });
});