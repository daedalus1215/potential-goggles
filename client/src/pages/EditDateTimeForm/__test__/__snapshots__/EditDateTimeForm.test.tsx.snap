// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/pages/tasks/TaskForm/dateTimeDetail/EditDateTimeForm/__test__/EditDateTimeForm.test.tsx EditDateTimeForm should render 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        data-testid="TopBar"
      >
        (
        <div
          data-testid="BackButton"
        >
          /task/randoTaskId/date-time
        </div>
        )
      </div>
      <div
        class="editDateTimeForm"
      >
        <div
          data-testid="Form"
        >
          (
          <h2
            class="h2"
          >
            Edit Date Time
          </h2>
          <input
            hidden=""
            name="taskId"
            readonly=""
            value="randoTaskId"
          />
          <input
            hidden=""
            name="id"
            readonly=""
            value=""
          />
          <input
            class="input"
            name="date"
            pattern="\\d{4}-[0-1]\\d-[0-3]\\dT[0-2]\\d:[0-5]\\d:[0-6]\\d.\\d{3}[A-Z]"
            type="text"
            value=""
          />
          <input
            class="input"
            name="minutes"
            type="number"
            value="123"
          />
          <div
            data-testid="SaveButton"
          >
            SaveButton
          </div>
          )
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      data-testid="TopBar"
    >
      (
      <div
        data-testid="BackButton"
      >
        /task/randoTaskId/date-time
      </div>
      )
    </div>
    <div
      class="editDateTimeForm"
    >
      <div
        data-testid="Form"
      >
        (
        <h2
          class="h2"
        >
          Edit Date Time
        </h2>
        <input
          hidden=""
          name="taskId"
          readonly=""
          value="randoTaskId"
        />
        <input
          hidden=""
          name="id"
          readonly=""
          value=""
        />
        <input
          class="input"
          name="date"
          pattern="\\d{4}-[0-1]\\d-[0-3]\\dT[0-2]\\d:[0-5]\\d:[0-6]\\d.\\d{3}[A-Z]"
          type="text"
          value=""
        />
        <input
          class="input"
          name="minutes"
          type="number"
          value="123"
        />
        <div
          data-testid="SaveButton"
        >
          SaveButton
        </div>
        )
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
